ansible-ipi-install/roles/installer/tasks/24_rhcos_image_cache.yml

```diff
@@ -0,0 +1,87 @@
+---
+# If we're executing these plays, then:
+# 1. cache_enabled is true
+# 2. Either one or both bootstraposimage/clusterosimage variables are unset
+
+- name: Create {{ provision_cache_store }} on provisioning host
+  file:
+    path: "{{ provision_cache_store }}"
+    state: directory
+    owner: "{{ ansible_user }}"
+    group: "{{ ansible_user }}"
+    mode: '0755'
+
+# rhcos_json fact already set in 23_rhcos_image_paths.yaml
+- name: Set facts for RHCOS_QEMU_SHA256 and RHCOS_SHA256
+  set_fact:
+    rhcos_qemu_sha256: '{{ rhcos_json.json | json_query(''images.qemu."uncompressed-sha256"'') }}'
+    rhcos_sha256: "{{ rhcos_json.json | json_query('images.openstack.sha256') }}"
+
+- name: Check if {{ rhcos_qemu_uri }} is already in cache
+  stat: 
+    path: "{{ provision_cache_store }}{{ rhcos_qemu_uri }}"
+    checksum_algorithm: sha256
+    get_checksum: yes
+  register: rhcos_qemu_stat
+  when: bootstraposimage is not defined or bootstraposimage|length < 1
+
+- name: Download {{ rhcos_qemu_uri }} for cache
+  get_url:
+    url:  "{{ rhcos_path }}{{ rhcos_qemu_uri }}"
+    dest: "{{ provision_cache_store }}{{ rhcos_qemu_uri }}"
+    owner: "{{ ansible_user }}"
+    group: "{{ ansible_user }}"
+    mode: '0644'
+    timeout: 3600
+  when: (bootstraposimage is not defined or bootstraposimage|length < 1) and (rhcos_qemu_stat.stat.exists == false or rhcos_qemu_stat.stat.checksum != rhcos_qemu_sha256)
+
+- name: Check if {{ rhcos_uri }} is already in cache
+  stat: 
+    path: "{{ provision_cache_store }}{{ rhcos_uri }}"
+    checksum_algorithm: sha256
+    get_checksum: yes
+  register: rhcos_stat
+  when: clusterosimage is not defined or clusterosimage|length < 1
+
+- name: Download {{ rhcos_uri }} for cache
+  get_url:
+    url:  "{{ rhcos_path }}{{ rhcos_uri }}"
+    dest: "{{ provision_cache_store }}{{ rhcos_uri }}"
+    owner: "{{ ansible_user }}"
+    group: "{{ ansible_user }}"
+    mode: '0644'
+    timeout: 3600
+  when: (clusterosimage is not defined or clusterosimage|length < 1) and (rhcos_stat.stat.exists == false or rhcos_stat.stat.checksum != rhcos_sha256)
+
+- name: Set bootstrap image URL override if not provided by the user
+  set_fact:
+    bootstraposimage: "http://172.22.0.1/{{ rhcos_qemu_uri }}?sha256={{ rhcos_qemu_sha256 }}"
+  when: bootstraposimage is not defined or bootstraposimage|length < 1
+
+- name: Set cluster image URL override if not provided by the user
+  set_fact:
+    clusterosimage: "http://172.22.0.1/{{ rhcos_uri }}?sha256={{ rhcos_sha256 }}"
```

Comment:
lets not use hardcoded 172.22.0.1, instead lets use
`{{ ansible_provisioning.ipv4.address }}`